compiler/resolution/implementForallIntents2.cpp

```diff
@@ -0,0 +1,1097 @@
+  USR_PRINT(enclPar, "the second enclosing parallel loop is here");
+  USR_PRINT(ctx.forall, "the forall loop that invokes the iterator is here");
+
+  // If we continue, we may get a const violation on x*_reduceShadowVar.
+  USR_STOP();
+}
+
+
+/////////// extendYieldNew, simple case ///////////
+
+static Symbol* tupcomForSerialYield(FIcontext& ctx, int ix)
+{
+  FIinfo& fii = ctx.fiInfos[ix];
+
+  switch (fii.svSymbol->intent) {
+    case  TFI_DEFAULT:
+    {
+      INT_ASSERT(false); // it is better to have a concrete intent here
+      return NULL;
+    }
+
+    case TFI_CONST:
+    case TFI_IN:
+    case TFI_CONST_IN:
+    case TFI_REF:
+    case TFI_CONST_REF:
+    {
+      Symbol* component = fii.fiFormal;
+      // When the type is dtUnknown, we still need to ensure yield by ref.
+      // Ex. have makeAddrOfIfNeeded() insert a PRIM_SET_REFERENCE.
+      INT_ASSERT(component->type != dtUnknown &&
+                 component->type != dtAny);
+      return component;
+    }
+
+    case TFI_REDUCE:
+    {
+      ensureCurrentReduceOpForReduceIntent(ctx, fii, ix);
+      return shadowVarForReduceIntent(ctx, fii, ix);
+    }
+  }
+
+  INT_ASSERT(false); // garbage intent
+  return NULL;
+}
+
+//
+// Switch from yielding a user value to yielding the tuple
+//   (user value, shadow variable1, shadow variable2...)
+//
+// 'efs' is the enclosing ForallStmt if any, otherwise NULL.
+//
+static void extendYieldNew(FIcontext& ctx, CallExpr* yieldCall,
+                           ForallStmt* efs)
+{
+  // Make a tuple that includes the extra args.
+  Expr*      origRetArg = yieldCall->get(1)->remove();
+  VarSymbol* newOrigRet = localizeYieldForExtendLeader(origRetArg, yieldCall);
+  CallExpr*  buildTuple = new CallExpr("_build_tuple_always_allow_ref",
+                                       newOrigRet);
+  int numSVars   = ctx.numShadowVars();
+
+  if (efs != NULL)
+    checkFor2ndEnclosingParallelLoop(ctx, yieldCall, efs);
+
+  for (int ix = 0; ix < numSVars; ix++) {
+    Symbol* tupComponent = (efs == NULL) ?
+      tupcomForSerialYield(ctx, ix) :
+      tupcomForYieldInForall(ctx, efs, ix);
+
+    // shadow variables are always yielded by reference
+    tupComponent = makeAddrOfIfNeeded(tupComponent, yieldCall);
+    buildTuple->insertAtTail(new SymExpr(tupComponent));
+  }
+
+  yieldCall->insertBefore("'move'(%S,%E)", ctx.retSym, buildTuple);
+  yieldCall->insertAtTail(new SymExpr(ctx.retSym));
+}
+
+
+/////////// propagateExtraLeaderArgsNew ///////////
+
+static ArgSymbol* newExtraFormal(FIcontext& ctx, FIinfo& fii, int ix,
+                                 Symbol* eActual)
+{
+    bool     isReduce   = fii.svSymbol->isReduce();
+    const char* eName   =
+      isReduce ? astrArg(ix, "reduceParent") :
+        ctx.nested ? eActual->name :
+          !strcmp(eActual->name, "_tuple_expand_tmp_") ?
+            astrArg(ix, "tet") :
+              astrArg(ix, eActual->name); // uniquify user arg name
```

Comment:
Would this be clearer as a series of conditionals? Not sure we're getting much by using the ternary operator here.